src/main/kotlin/org/rust/lang/core/macros/ExpandedMacroStorage.kt

```diff
@@ -0,0 +1,626 @@
+                    val sourceFiles = ArrayList<SourceFile>(sourceFilesSize)
+
+                    val storage = ExpandedMacroStorage(project)
+                    for (i in 0 until sourceFilesSize) {
+                        sourceFiles += SourceFile.readFrom(data, storage) ?: continue
+                    }
+                    storage.deserialize(sourceFiles)
+                    storage
+                }
+            } catch (e: FileNotFoundException) {
+                null
+            } catch (e: java.nio.file.NoSuchFileException) {
+                null
+            } catch (e: Exception) {
+                LOG.warn(e)
+                null
+            }
+        }
+
+        fun saveStorage(storage: ExpandedMacroStorage, dataFile: Path) {
+            Files.createDirectories(dataFile.parent)
+            DataOutputStream(DeflaterOutputStream(Files.newOutputStream(dataFile))).use { data ->
+                data.writeInt(STORAGE_VERSION)
+                data.writeInt(RsFileStub.Type.stubVersion)
+
+                data.writeInt(storage.sourceFiles.size)
+                storage.sourceFiles.values.forEach { it.writeTo(data) }
+            }
+        }
+    }
+}
+
+class SourceFile(
+    val storage: ExpandedMacroStorage,
+    val file: VirtualFile,
+    val depth: Int,
+    private var fresh: Boolean = true,
+    private var modificationStamp: Long = -1,
+    private val infos: MutableList<ExpandedMacroInfo> = mutableListOf()
+) {
+    private var cachedPsi: SoftReference<RsFile>? = null
+
+    val fileUrl: String = file.url
+
+    val project: Project
+        get() = storage.project
+
+    private val rootSourceFile: SourceFile by lazy(LazyThreadSafetyMode.PUBLICATION) {
+        findRootSourceFile()
+    }
+
+    // Should not be lazy b/c target/pkg/origin can be changed
+    private val isBelongToWorkspace: Boolean
+        get() = rootSourceFile.loadPsi()?.containingCargoTarget?.pkg?.origin == PackageOrigin.WORKSPACE
+
+    @Synchronized
+    fun getExpansionUrls(): List<Pair<String, ExpandedMacroInfo>> =
+        infos.mapNotNull { info -> info.expansionFileUrl?.let { it to info } }
+
+    @Synchronized
+    fun addInfo(info: ExpandedMacroInfo) {
+        check(!fresh)
+        infos.add(info)
+        detectDuplicates()
+    }
+
+    @Synchronized
+    fun removeInfo(info: ExpandedMacroInfo) {
+        check(!fresh)
+        infos.remove(info)
+    }
+
+    fun isEmpty(): Boolean =
+        infos.isEmpty()
+
+    fun extract(workspaceOnly: Boolean): List<Pipeline.Stage1> {
+        if (workspaceOnly && !isBelongToWorkspace) {
+            // very hot path; the condition should be as fast as possible
+            return emptyList()
+        }
+
+        synchronized(this) {
+            rebind()
+            return infos.map { info ->
+                info.makePipeline()
+            }
+        }
+    }
+
+    @Synchronized
+    fun bindCalls(calls: List<RsMacroCall>): List<Pipeline.Stage1> {
```

Comment:
This function is called only in one place (in `makeExpansionTask`); however, its return value is not used there. Maybe it should not create that list (since it's not used) and not create `Stage1`? This way, `Stage1`'s constructor may become more strict (accept `ExpandedMacroInfo` instead of `ExpandedMacroInfo?`)

Reply:
Removing this does not affect the passing of the tests

Reply:
:+1: :+1: :+1: